### PR TITLE
Add step dataset-filetypes

### PR DESCRIPTION
**Pull request by severo**
needed for https://github.com/huggingface/dataset-viewer/issues/2898#issuecomment-2162460762

### CODE REVIEW

File: services/worker/src/worker/job_runners/dataset/filetypes.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    try:
+        info = HfApi(endpoint=hf_endpoint, token=hf_token).dataset_info(dataset)
+    except RepositoryNotFoundError as err:
+        raise DatasetNotFoundError(f"Cannot get the dataset info for {dataset=}") from err
+
+    # get file types count
+    filetypes = get_filetypes(info.siblings)
+
+    # look into the zip archives to get the file types
+    SUPPORTED_ARCHIVE_EXTENSIONS = [".zip"]
```
Comment (severo) [inline services/worker/src/worker/job_runners/dataset/filetypes.py:89]: Should we add other extensions?

Comment (lhoestq) [reply services/worker/src/worker/job_runners/dataset/filetypes.py:89]: ZIP only is fine

We don't support RAR or 7Z in `datasets`

**Review by lhoestq — COMMENTED** [review]
Nice ! just some minor comments:

File: services/worker/src/worker/utils.py [hunk 2]
```diff
@@ -238,3 +239,17 @@ def raise_if_long_column_name(features: Optional[Features]) -> None:
...
+    # special cases we find in datasets (gz?dl=1 -> gz, txt_1 -> txt, txt-00000-of-00100-> txt)
+    # https://github.com/huggingface/datasets/blob/af3acfdfcf76bb980dbac871540e30c2cade0cf9/src/datasets/utils/file_utils.py#L795
+    if clean:
+        for symb in "?-_":
+            extension = extension.split(symb)[0]
+    if recursive and extension in [".gz", ".bz2", ".xz", ".zst", ".7z"]:
+        base_extension = get_file_extension(base, recursive=False, clean=False)
+        if base_extension == ".tar":
+            return base_extension + extension
+    return extension
```
[the remark this answers is not in the document]
Comment (severo) [reply services/worker/src/worker/utils.py:255]: indeed, I only considered .tar.gz. TBH, I don't know what we should do

File: services/worker/tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        ("file.tar", ".tar"),
+        ("file.nottar.gz", ".gz"),
+        # clean suffixes
+        ("file.csv?dl=1", ".csv"),
+        ("file.csv_1", ".csv"),
+        ("file.csv-00000-of-00001", ".csv"),
+        # ignore paths
+        ("path/to/file.csv", ".csv"),
+        (".path/to.some/file.csv", ".csv"),
+        ("path/to/.gitignore", ""),
```
[the remark this answers is not in the document]
Comment (severo) [reply services/worker/tests/test_utils.py:36]: hmmm, I'm not sure we want to do this.

I think we want to return two entries in this case: one for .gz, and one for jsonl, with `archived_in: "gz"`, no?

File: services/worker/src/worker/utils.py [hunk 4]
```diff
@@ -238,3 +239,17 @@ def raise_if_long_column_name(features: Optional[Features]) -> None:
...
+
+
+def get_file_extension(filename: str, recursive: bool = True, clean: bool = True) -> str:
+    [base, extension] = os.path.splitext(filename)
+    # special cases we find in datasets (gz?dl=1 -> gz, txt_1 -> txt, txt-00000-of-00100-> txt)
+    # https://github.com/huggingface/datasets/blob/af3acfdfcf76bb980dbac871540e30c2cade0cf9/src/datasets/utils/file_utils.py#L795
+    if clean:
+        for symb in "?-_":
+            extension = extension.split(symb)[0]
+    if recursive and extension in [".gz", ".bz2", ".xz", ".zst", ".7z"]:
```
[the remark this answers is not in the document]
Comment (severo) [reply services/worker/src/worker/utils.py]: OK. The change is that we add `.lz4` and drop `.7z`. Is it OK?

File: services/worker/src/worker/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (severo) [reply services/worker/src/worker/utils.py]: OK, I now count both the `.gz` file and the `.jsonl` uncompressed file

File: services/worker/tests/job_runners/dataset/test_filetypes.py [hunk 5]
```diff
@@ -39,21 +39,24 @@
...
             ],
             [
                 Filetype(extension=".txt", count=2),
                 Filetype(extension=".csv", count=1),
                 Filetype(extension=".tar", count=1),
-                Filetype(extension=".tar.gz", count=4),
-                Filetype(extension=".gz", count=1),
+                Filetype(extension=".gz", count=6),
+                Filetype(extension=".tar", count=4, archived_in=".gz"),
+                Filetype(extension=".json", count=1, archived_in=".gz"),
```
[the remark this answers is not in the document]
Comment (severo) [reply services/worker/tests/job_runners/dataset/test_filetypes.py:51]: hmmm, ok

[the remark this answers is not in the document]
Comment (severo) [reply services/worker/tests/job_runners/dataset/test_filetypes.py]: Done. You're right, it's better

**Review by lhoestq — APPROVED** [review]
lgtm !

**Pull request merged**